Initial commit of tcp provider. It is currently non functional

vkrishna opened the pull request:
Signed-off-by: Venkata Krishna Nimmagadda venkata.krishna.nimmagadda@intel.com

jsquyres commented:
What is the difference between this provider and the `sockets` provider?  Is this provider going to eventually obsolete the `sockets` provider?

shefty commented:
The hope is that this replaces the sockets provider, with more maintainable code.  I.e. greatly increased use of shared code.  (This will depend on achieving similar performance and scalability.)

The sockets provider is designed around supporting rdm ep's, with msg and dgram support implemented over rdm.  This provider will be designed around msg ep's, which is more natural for TCP sockets.  The rxm provider will expand that support to rdm.  Similarly, the udp provider replaces the socket dgram ep support.

If rxm + tcp cannot be an adequate replacement, then we will keep the socket provider, but limit it to rdm ep only.

shefty commented:
This is for post 1.5

shefty commented:
Please add signed-off-by lines to all commit messages.

vkrishna commented:
signed and pushed all commits. Thanks

vkrishna commented:
Fixed few issues in implementation around Provider version, passive endpoint setup and eq reporting

shefty commented:
Please respond when this PR is in a state for reviewing for possible upstreaming.  It doesn't need to work completely, or even at all really, as long as it doesn't break anything and what is there is a reasonable work in progress.

vkrishna commented:
Sean
As of now, It does not break anything and at a good point for upstreaming.

shefty commented on prov/tcp/src/tcpx_attr.c:
```diff
@@ -0,0 +1,96 @@
+	.cq_cnt = 256,
+	.ep_cnt = 1024,
+	.tx_ctx_cnt = 256,
+	.rx_ctx_cnt = 256,
```
tc_ctx_cnt and rx_ctx_cnt should match the ep_cnt.  For scaling, we'll want these values higher.  The ep_cnt will limit the number of peers that we can connect to.  This will ultimately be limited by the number of fd's that the app can open.  I'm not sure if you can dynamically get the value, but maybe bump this to 4k or 8k.  That will be higher than the fd limit by default.

shefty commented on prov/tcp/src/tcpx_attr.c:
```diff
@@ -0,0 +1,96 @@
+};
+
+struct fi_fabric_attr tcpx_fabric_attr = {
+	.name = "TCP-IP",
```
See sock_get_fabric_name().  For tcp the fabric name is best chosen as the subnet mask.  You may want to match the same semantics, even moving the sock_get_fabric_name() and sock_get_domain_name() calls into common util code.

shefty commented on prov/tcp/src/tcpx_attr.c:
```diff
@@ -0,0 +1,96 @@
+	.data_progress = FI_PROGRESS_AUTO,
+	.resource_mgmt = FI_RM_ENABLED,
+	.av_type = FI_AV_UNSPEC,
+	.mr_mode = FI_MR_UNSPEC,
```
The mr_mode was changed from an enum to flags.  Please set to 0.

shefty commented on prov/tcp/src/tcpx_init.c:
```diff
@@ -0,0 +1,68 @@
+struct fi_provider tcpx_prov = {
+	.name = "tcp",
+	.version = FI_VERSION(TCPX_MAJOR_VERSION,TCPX_MINOR_VERSION),
+	.fi_version = FI_VERSION(FI_MAJOR_VERSION,FI_MINOR_VERSION),
```
Don't use the FI_MAJOR_VERSION, FI_MINOR_VERSION defines.  Hard-code the actual version that the provider is coded to, e.g. 1, 5.  This ensures that if the version bumps to 1.6, that the provider does NOT report that it supports 1.6 features until it has actually been updated.

shefty commented on prov/tcp/src/tcpx_init.c:
```diff
@@ -0,0 +1,68 @@
+			 struct fi_info **info)
+{
+	return util_getinfo(&tcpx_util_prov, version, node, service, flags,
+			    hints, info);
```
This will need more work in order to report source addressing, but that can come later.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		    eq_entry, sizeof(*eq_entry), 0);
+
+
+
```
nit - extra blank lines

Need to check for eq_write failures

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+	if (!eq_entry) {
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "cannot allocate memory\n");
+	  goto err;
```
indentation is off

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+	fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_SHUTDOWN,
+		    eq_entry, sizeof(*eq_entry), 0);
+
+
```
extra blank lines -- need to check for eq write failure

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		  "connect failed : %s\n", strerror(errno));
+	  goto err1;
+	}
+	fi_fd_nonblock(handle->conn_fd);
```
May want to merge this with tcpx_set_sockopts() -> tcpx_setup_socket().

shefty replied:
This is using a blocking connect() call.  This will need to convert to non-blocking for better asynchronous processing.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+
+	  FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+		  "connect failed : %s\n", strerror(errno));
+	  goto err1;
```
indentation is off -- make sure using tab stops set to 8 characters, and not converting to spaces.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+
+	/* report the error to eq */
+	fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_SHUTDOWN,
+		    eq_entry, sizeof(*eq_entry), 0);
```
Err1 path frees eq_entry, which is then written to the eq.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+	if (!addr)
+		return -FI_EINVAL;
+
+	handle->conn_fd = socket(AF_INET, SOCK_STREAM, 0);
```
We should be able to allocate the socket during EP creation.  This will allow us to bind the source address prior to the user calling connect.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"failed to create connect thread\n");
+		goto err;
+	}
```
We should be able to start a non-blocking connect here, with a cm thread finishing the connection.

nit: extra blank line below

vkrishna replied:
Does it mean it is okay to have this thread per ep or should this functionality be moved to domain's/fabric's dedicated thread?

shefty replied:
I would try to limit the code to a single thread per provider, or at least fabric.  Recent changes to the socket provider were made to reduce the number of threads.  We don't want to start out with more, just to rework the code to reduce it later.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		goto out;
+	}
+
+	tcpx_ep->handle->serv_addr = serv_addr;
```
I don't think we need to store the server address.  I haven't checked where this is used, but from OFI's perspective, I don't think it's necessary.  getpeername() could be used to obtain the peer address if it's needed.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		eq_entry->info = &tcpx_ep->info;
+
+		/* report FI_CONNECTED to eq  */
+		fi_eq_write(&tcpx_ep->util_ep.eq->eq_fid, FI_CONNECTED, eq_entry,
```
Need to check for failure

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+	struct fi_eq_cm_entry *eq_entry;
+	struct tcpx_ep *tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+
+	if (tcpx_ep->handle != NULL && tcpx_ep->handle->conn_fd != 9) {
```
magic number 9?

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+			    sizeof(*eq_entry), 0);
+		return 0;
+	}
+	return -FI_EINVAL;
```
Reverse the above if statement to check for a failure and abort immediately, rather than having the entire successful code path indented inside of a if statement.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+	struct tcpx_ep *ep;
+
+	ep = container_of(fid, struct tcpx_ep, util_ep.ep_fid.fid);
+	if (ofi_atomic_get32(&ep->ref)) {
```
What takes a reference on the ep?

vkrishna replied:
Nothing. I will remove it.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		tcpx_ep->util_ep.eq = eq;
+		ofi_atomic_inc32(&eq->ref);
+
+
```
extra lines

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+					      &tcpx_ep->util_ep.ep_fid.fid);
+			if (ret)
+				return ret;
+
```
extra line

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+					      &tcpx_ep->util_ep.ep_fid.fid);
+			if (ret)
+				return ret;
+
```
extra line

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		return -FI_ENOSYS;
+	}
+	return FI_SUCCESS;
+}
```
This function can probably be dropped in favor of using ofi_ep_bind().

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		handle = container_of(info->handle,struct tcpx_conn_handle,
+				      handle);
+		ep->handle = handle;
+		fi_fd_nonblock(handle->conn_fd);
```
I would look at moving the socket directly into the ep structure, and not a substructure.  The connection request handle can be freed immediately after the connection completes.

We should allocate the socket in this call, and configure it appropriately.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+	}
+	pep->eq = eq;
+	return 0;
+}
```
See about moving all or portions of the pep close and bind calls into common code, with a util_pep structure.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+	poll_fds[1].fd = pep->cm.signal_fds[1];
+	poll_fds[0].events = poll_fds[1].events = POLLIN;
+
+	while (*((volatile int *) &pep->cm.do_listen)) {
```
Volatile is almost always wrong unless there is hardware (e.g. a NIC) capable of changing the memory.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+					       "Invalid signal\n");
+				continue;
+			}
+
```
extra line

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+			}
+
+		} else {
+			break;
```
add a warning message that poll has failed.  We shouldn't be here.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		if (!cm_entry) {
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"cannot allocate memory\n");
+			goto err;
```
This is unlikely to happen, but if it does, close the socket and continue processing.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		/* report the conn req to the associated eq */
+		if (fi_eq_write(&pep->eq->eq_fid, FI_CONNREQ, cm_entry, entry_sz, 0) < 0)
+			FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+				"Error in writing to EQ\n");
```
On an error we need to close the socket and simply continue.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+			ofi_close_socket(_pep->cm.sock);
+			_pep->cm.sock = -1;
+		}
+	}
```
We should allocate the passive socket when creating the endpoint.  We can bind() it to the desired address at that time.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+			"failed to create cm thread\n");
+		goto out;
+	}
```
You mentioned this, but we will want a single thread per fabric or domain, rather than per PEP.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+		return -FI_EIO;
+	}
+
+	if (listen(_pep->cm.sock, TCPX_MAX_SOCK_REQS)) {
```
Use the default connection backlog.

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+			freeaddrinfo(result);
+		}
+
+	} else {
```
Move !info check to the top of the function and abort early.  This avoids unnecessary code indentation

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+ 	ret = fi_fd_nonblock(_pep->cm.signal_fds[1]);
+	if (ret) {
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"fi_fd_nonblock failed\n");
+	}
```
Use fi_signal

shefty commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,769 @@
+			memcpy(&_pep->src_addr, result->ai_addr,
+			       result->ai_addrlen);
+			freeaddrinfo(result);
+		}
```
We should be able to allocate the socket here and bind it to the src_addr in fi_info.

vkrishna commented:
@shefty Thanks for review comments.

Here is the summary of changes in my latest commit:
- changed tcpx message size, ep_cnt and context counts aptly
- added support for ipv6
- Switched to using signaling abstraction in fi_signal.h
- cleaned tcpx structures (except the need for fi_info)
- consolidated threads and now we have only one thread in fabric
--- Now Fabric has a thread which handles both PEP listening and EP connects.
- removed unnecessary blanks lines and added blank line after arguments in tcpx_ep.c
- moved socket creations to initialization routines in ep and pep cases.

I will be working on remaining comments and update.

a-ilango commented on prov/tcp/src/tcpx_ep.c:
```diff
@@ -0,0 +1,620 @@
+		addrlen = INET_ADDRSTRLEN;
+		break;
+	case AF_INET6:
+		addrlen = INET6_ADDRSTRLEN;
```
addrlen arg to connect( ) should be size of the struct sockaddr_in/sockaddr_in6 rather than size of string representation of the address.

a-ilango commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,331 @@
+						"getsockopt failed\n");
+				if (status == 0) {
+
+					eq_entry = calloc(1, sizeof(*eq_entry));
```
This can be a local variable instead of being dynamically allocated since this is used only in this function.

shefty commented on prov/tcp/src/tcpx.h:
```diff
@@ -0,0 +1,137 @@
+struct tcpx_cm_entry {
+	int sock;
+	int sock_fd_closed;
+};
```
This can probably be merged directly into tcpx_pep

shefty commented on prov/tcp/src/tcpx.h:
```diff
@@ -0,0 +1,137 @@
+
+struct poll_fd_info {
+	fid_t fid;
+	int sock_fd;
```
The fid should be castable to the correct structure to get the fd.

vkrishna replied:
If I understand correctly, what you are saying here is to use fid to get the ep or pep structure and extract fd from it rather than having an extra variable "sock_fd". Is that right?

shefty commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,319 @@
+				if (ret < 0)
+					FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+						"getsockopt failed\n");
+				if (status == 0) {
```
status can be used uninitialized in this function.  I would drop this check.  If getsockopt fails or status is an error, we should report a connection failure to the EQ.

shefty commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,319 @@
+					FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+						"Error in writing to EQ\n");
+					free(handle);
+					ofi_close_socket(handle->conn_fd);
```
use of handle after freeing it

shefty commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,319 @@
+
+	for (iter = 0; iter < *nfds ; iter++) {
+		switch (fd_info[iter].fid->fclass) {
+		case FI_CLASS_EP:
```
You can check for revent first and remove the switch statement.  POLLOUT only applies to EPs and POLLIN only to PEPs.  You could assert that in the code.

shefty commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,319 @@
+							"Error in writing to EQ\n");
+					}
+				}
+				fd_list_delete_item(iter, poll_fds,fd_info, nfds);
```
This is modifying the array while you are attempting to iterate over it.  This can result in the item at the end of the list, which ends up being moved to this location, being skipped in processing.  There's not a great fix for this.  You can place the item to remove on a cleanup list and perform that after the array has been processed.

vkrishna replied:
Is it okay to just decrement the iter after remove?

shefty commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,319 @@
+					struct poll_fd_info,
+					entry));
+		if (*nfds < 1024) {
+			for (i =0 ; i < *nfds ; i++) {
```
Maybe add a function to find the index given a socket value, or have a remove from poll set function do the search.

shefty commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,319 @@
+			default:
+				FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,
+					"invalid fd\n");
+			}
```
Move into a function to add a new item into the poll set.

shefty commented:
> +struct poll_fd_info {
> +	fid_t fid;
> +	int sock_fd;
> 
> If I understand correctly, what you are saying here is to use fid to
> get the ep or pep structure and extract fd from it rather than having
> an extra variable "sock_fd". Is that right?

correct

shefty commented:
> Is it okay to just decrement the iter after remove?

That will work, but would be fragile.  The decrement is dependent on how the remove function is implemented, and that the index won't be used again in the loop until it is incremented.  An alternative solution would be preferable if it can be cleaner.

vkrishna commented:
1) I believe renaming the function to "swap_and_del_last_item" and going through array index from high to low (instead of low to high) could help. 
OR
2) Adjust the array before calling poll

shefty commented:
> 1.	I believe renaming the function to "swap_and_del_last_item" and
> going through array index from high to low (instead of low to high)
> could help.

This sounds promising.

vkrishna commented on prov/tcp/src/tcpx_attr.c:
```diff
@@ -0,0 +1,96 @@
+	.ep_attr = &tcpx_ep_attr,
+	.domain_attr = &tcpx_domain_attr,
+	.fabric_attr = &tcpx_fabric_attr
+};
```
Yes. Made them static

vkrishna commented on prov/tcp/src/tcpx_fabric.c:
```diff
@@ -0,0 +1,319 @@
+		fd_entry = tcpx_fabric->fd_list.next;
+		fd_info_entry = (container_of(fd_entry,
+					struct poll_fd_info,
+					entry));
```
Done

vkrishna commented:
Added support for setting fabric and domain names in util code.
Made changes as per review.

shefty commented:
Please see travis CI failures.  There are also now conflicts in Makefile.am (likely merging of shared memory provider).  Please resolve the CI failures, and I'll start on another round of review with the goal to start merging the code into mainline.

vkrishna commented:
I have found couple of issues and fixed and merged latest changes. Right now travis is failing for these tests

fi_dgram
fi_dgram_pingpong
fi_dgram_waitset

All these tests get fi_info for FI_EP_DGRAM endpoint type which includes shm provider. These tests are making fi_getinfo calls with node parameter as NULL . In case of shm provider this is causing seg fault since ofi_get_addr is calling strdup on node which is NULL .

shefty commented:
@vkrishna - please open an issue regarding the failures and tag @aingerson on them -- thanks

vkrishna commented:
Sorry for confusion. That is not the issue.

vkrishna commented:
Added a callback function in util_getinfo to have provider specific processing of fabric and domain names.

shefty commented:
Replaced by #3507
